resources/api/resource.py

```diff
@@ -339,5 +370,33 @@ def get_queryset(self):
         else:
             return self.queryset.filter(public=True)
 
+    @detail_route(methods=['post'])
+    def favorite(self, request, pk=None):
+        resource = self.get_object()
+        user = request.user
+
+        if not resource.is_admin(user):
+            raise PermissionDenied()
+
+        if user.favorite_resources.filter(id=resource.id).exists():
+            return response.Response(status=status.HTTP_304_NOT_MODIFIED)
+        else:
+            user.favorite_resources.add(resource)
+            return response.Response(status=status.HTTP_201_CREATED)
+
+    @detail_route(methods=['post'])
+    def unfavorite(self, request, pk=None):
```

Comment:
The methods share so much code that it would be better to refactor them to call a common setter function.